<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>A Configurable shiny Portal for Sharing Analysis of Molecular Expression Data • shinyExprPortal</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="A Configurable shiny Portal for Sharing Analysis of Molecular Expression Data">
<meta name="description" content="Enables deploying configuration file-based shiny apps with minimal programming for interactive exploration and analysis showcase of molecular expression data. For exploration, supports visualization of correlations between rows of an expression matrix and a table of observations, such as clinical measures, and comparison of changes in expression over time. For showcase, enables visualizing the results of differential expression from package such as limma, co-expression modules from WGCNA and lower dimensional projections.">
<meta property="og:description" content="Enables deploying configuration file-based shiny apps with minimal programming for interactive exploration and analysis showcase of molecular expression data. For exploration, supports visualization of correlations between rows of an expression matrix and a table of observations, such as clinical measures, and comparison of changes in expression over time. For showcase, enables visualizing the results of differential expression from package such as limma, co-expression modules from WGCNA and lower dimensional projections.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">shinyExprPortal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/shinyExprPortal.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/config.html">Full Configuration Guide</a></li>
    <li><a class="dropdown-item" href="articles/customization.html">Portal UI and Code Customization</a></li>
    <li><a class="dropdown-item" href="articles/dataprep.html">Data Preparation Guide</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/C4TB/shinyExprPortal/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="shinyexprportal">shinyExprPortal<a class="anchor" aria-label="anchor" href="#shinyexprportal"></a>
</h1></div>
<!-- badges: start -->

<p><em>shinyExprPortal</em> is a configurable Shiny portal for the visual exploration and analysis showcase of molecular expression data and phenotype data, such as observed clinical measures. The package is designed for deploying portals using a text-based configuration file with minimal programming. The portal includes modules for exploring correlations between expression and measures, visualizing the results of differential expression analysis and showcasing results of other downstream analysis methods such as pathways and correlated networks. The interface is designed for gene expression, but the computational methods are compatible with any matrix with samples in columns and entities, such as proteins, in rows.</p>
<p><img src="reference/figures/example2.png"></p>
<p>For initial exploration, any expression matrix containing abundance or counts and data frames containing observed measures can be loaded in. For differential expression and other methods, the package is compatible with the outputs of packages such as limma and WCGNA. The correlation modules can be used early in the lifecycle of a bioinformatics project, while the other modules can be added as new downstream analysis results are produced. The use of the configuration file also means that the portal can be easily versioned.</p>
<p>A demo of the portal is available <a href="https://c4tb.github.io/ramap_demo/" class="external-link">here</a>.</p>
<div class="section level2">
<h2 id="how-to-start">How to start<a class="anchor" aria-label="anchor" href="#how-to-start"></a>
</h2>
<p>The package relies on other packages from CRAN and Bioconductor. To ensure the installation goes smoothly, install BiocManager by running:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"BiocManager)</span></span></code></pre></div>
<p>You can then install the latest version of the package using:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"shinyExprPortal"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can install the current development version using devtools (installing devtools first if necessary):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"C4TB/shinyExprPortal"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>If you only have one set of set of samples for a group of subjects, i.e. one sample per patient, you can check <a href="https://c4tb.github.io/shinyExprPortal/articles/shinyExprPortal.html"><code>vignette("shinyExprPortal")</code></a> for a quick start guide to set up the configuration file.</p>
<p>If your subjects have one more samples, e.g. collected over time or from different tissues, you should check the data preparation guide in <a href="https://c4tb.github.io/shinyExprPortal/articles/dataprep.html"><code>vignette("dataprep")</code></a> and the full configuration guide <a href="https://c4tb.github.io/shinyExprPortal/articles/config.html"><code>vignette("config")</code></a></p>
<p>Once you have completed either of the steps above, you can open and run the app.R file created to test the portal.</p>
</div>
<div class="section level2">
<h2 id="module-documentation">Module documentation<a class="anchor" aria-label="anchor" href="#module-documentation"></a>
</h2>
<p>The full configuration guide in <a href="https://c4tb.github.io/shinyExprPortal/articles/config.html"><code>vignette("config")</code></a> describes all the dependencies, file requirements, mandatory and optional settings for each of the modules currently supported by the package.</p>
</div>
<div class="section level2">
<h2 id="customization-and-extending-the-package">Customization and extending the package<a class="anchor" aria-label="anchor" href="#customization-and-extending-the-package"></a>
</h2>
<p><a href="https://c4tb.github.io/shinyExprPortal/articles/customization.html"><code>vignette("customization")</code></a> describes how to customize the visual style of the portal and how to add new functionality through programming new modules. The package supports customization via the configuration file at a global level using <a href="https://CRAN.R-project.org/package=bslib" class="external-link">bslib</a>. Some modules also support the customization of the colors used in the visualizations.</p>
<p>The package is also compatible with new modules that can be added without modifying the source code of the portal or the package. New functionality can be developed and tested without having to clone the repository or recompile the package. We are also open for suggestions of new modules, please use the GitHub issues page for that.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=shinyExprPortal" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/C4TB/shinyExprPortal/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/C4TB/shinyExprPortal/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing shinyExprPortal</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Rafael Henkin <br><small class="roles"> Maintainer, author, copyright holder </small> <a href="https://orcid.org/0000-0002-5511-5230" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/C4TB/shinyExprPortal/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/C4TB/shinyExprPortal/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://CRAN.R-project.org/package=shinyExprPortal" class="external-link"><img src="https://www.r-pkg.org/badges/version/shinyExprPortal" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Rafael Henkin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
