<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="shinyExprPortal">
<title>Data Preparation Guide • shinyExprPortal</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data Preparation Guide">
<meta property="og:description" content="shinyExprPortal">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">shinyExprPortal</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0.0000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/clinvisx.html">Quick Start Guide for shinyExprPortal</a>
    <a class="dropdown-item" href="../articles/config.html">Full Configuration Guide</a>
    <a class="dropdown-item" href="../articles/customization.html">Portal UI and Code Customization</a>
    <a class="dropdown-item" href="../articles/dataprep.html">Data Preparation Guide</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Data Preparation Guide</h1>
            
      
      
      <div class="d-none name"><code>dataprep.Rmd</code></div>
    </div>

    
    
<p>The package was developed to support the most common data shapes and
file formats used in bioinformatics. This guide describes how each filed
used by the portal should be structured.</p>
<p><strong>Expression matrix</strong></p>
<p>The expected format for matrices contains sample identifiers in
columns and symbols in rows. The <strong>best</strong> format for the
portal is saving an R object of matrix type as an .rds file because it
will load faster, although it is also possible to save the matrix as a
CSV or TSV file. If using the latter, ensure that the first row (with
exception of the first column) contain the sample identifiers and that
the first column contain the transcript identifiers.</p>
<blockquote>
<p><strong>Saving an RDS file</strong>: to export a matrix as an rds
file, simply run:<br><code>saveRDS(matrix_object, "matrix_object.rds")</code></p>
</blockquote>
<p>The following is a valid example of an expression matrix:</p>
<table class="table">
<colgroup>
<col width="5%">
<col width="15%">
<col width="15%">
<col width="15%">
<col width="15%">
<col width="15%">
<col width="15%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">S1_01</th>
<th align="right">S1_02</th>
<th align="right">S2_01</th>
<th align="right">S2_02</th>
<th align="right">S3_01</th>
<th align="right">S3_02</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ABC</td>
<td align="right">-0.5241337</td>
<td align="right">1.0355301</td>
<td align="right">-1.5874156</td>
<td align="right">-0.8446914</td>
<td align="right">0.7121033</td>
<td align="right">0.7522620</td>
</tr>
<tr class="even">
<td align="left">BCD</td>
<td align="right">0.8711744</td>
<td align="right">-1.3920630</td>
<td align="right">-0.0808198</td>
<td align="right">-0.9204510</td>
<td align="right">1.1575474</td>
<td align="right">0.2955323</td>
</tr>
<tr class="odd">
<td align="left">CDE</td>
<td align="right">-0.5514590</td>
<td align="right">-1.5983642</td>
<td align="right">-0.2520080</td>
<td align="right">-0.0165644</td>
<td align="right">1.2594278</td>
<td align="right">0.6319722</td>
</tr>
<tr class="even">
<td align="left">DEF</td>
<td align="right">-0.1622565</td>
<td align="right">-0.2819614</td>
<td align="right">1.2852157</td>
<td align="right">-2.2808580</td>
<td align="right">-0.0088550</td>
<td align="right">1.2526449</td>
</tr>
<tr class="odd">
<td align="left">EFG</td>
<td align="right">-0.2875834</td>
<td align="right">-0.0952393</td>
<td align="right">0.8679439</td>
<td align="right">-0.2411127</td>
<td align="right">0.8931705</td>
<td align="right">-0.8716729</td>
</tr>
<tr class="even">
<td align="left">FGH</td>
<td align="right">0.0251335</td>
<td align="right">0.5005478</td>
<td align="right">1.0184627</td>
<td align="right">0.1937303</td>
<td align="right">0.4580773</td>
<td align="right">-0.5904114</td>
</tr>
<tr class="odd">
<td align="left">GHI</td>
<td align="right">-0.9086778</td>
<td align="right">2.2340152</td>
<td align="right">-0.6324829</td>
<td align="right">-0.4413873</td>
<td align="right">1.2573461</td>
<td align="right">0.3712248</td>
</tr>
<tr class="even">
<td align="left">HIJ</td>
<td align="right">-0.6921311</td>
<td align="right">-1.0550491</td>
<td align="right">0.8878756</td>
<td align="right">0.6780634</td>
<td align="right">0.3310022</td>
<td align="right">0.0073507</td>
</tr>
<tr class="odd">
<td align="left">IJK</td>
<td align="right">1.4220109</td>
<td align="right">-0.6713951</td>
<td align="right">-1.8607245</td>
<td align="right">-1.9963439</td>
<td align="right">0.7207575</td>
<td align="right">-0.6118567</td>
</tr>
<tr class="even">
<td align="left">JKL</td>
<td align="right">-1.2662481</td>
<td align="right">0.1053154</td>
<td align="right">-0.7174063</td>
<td align="right">-0.1526645</td>
<td align="right">-2.8115329</td>
<td align="right">0.5273308</td>
</tr>
</tbody>
</table>
<p>It is also possible to use the <a href="https://cran.r-project.org/web/packages/fst/index.html" class="external-link">{fst}</a>
and <a href="https://cran.r-project.org/web/packages/qs/index.html" class="external-link">{qs}</a>
packages for compressed data files. If using any of these packages, use
.fst or .qs as file extensions.</p>
<p><strong>Measures table</strong></p>
<p>The measures table follows the format of one row per subject (even if
they have more than one sample collected) with the measures across
columns. A data.frame or tibble object can be saved in an .rds file and
used, otherwise CSV or TSV files are also supported.</p>
<p>Measures collected over time should be represented in separate
columns, with the convention (enforced by default) of a time code as a
suffix for measure names, separated by underscore (<code>_</code>) –
this means that underscore cannot be used in long measure names as well.
For example, for disease activity collected over four time points, the
expected names are: diseaseActivity_Baseline, diseaseActivity_Week1,
diseaseActivity_Week2 and diseaseActivity_Week3. By default, it is
invalid to use a name such as Disease_Activity_Baseline. The time
separator can be modified in the configuration file by setting
<code>timesep</code> to the desired separator.</p>
<p>The following is a valid example of a measures table:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Patient_ID</th>
<th align="right">Platelets_m01</th>
<th align="right">Platelets_m02</th>
<th align="right">Age</th>
<th align="left">drugNaive</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">p01</td>
<td align="right">248.3360</td>
<td align="right">215.7336</td>
<td align="right">59</td>
<td align="left">Yes</td>
</tr>
<tr class="even">
<td align="left">p02</td>
<td align="right">162.0921</td>
<td align="right">158.7200</td>
<td align="right">54</td>
<td align="left">Yes</td>
</tr>
<tr class="odd">
<td align="left">p03</td>
<td align="right">238.9092</td>
<td align="right">152.7366</td>
<td align="right">60</td>
<td align="left">No</td>
</tr>
</tbody>
</table>
<p><strong>Lookup table</strong></p>
<p>For datasets where a subject has more than one sample (e.g. samples
over time, from different tissues or combinations thereof), a lookup
table should be constructed and saved as a data.frame or tibble format
in rds, CSV or TSV.</p>
<p>This table will map subject identifiers to sample identifiers, with
the expression matrix containing data for <em>all</em> samples in the
dataset. The table should also contain metadata that allows selection of
groups of samples. For example, if subjects and samples vary on time,
drug groups and tissues, the lookup table should have one column for
each category. The following is an example of such a table:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Sample_ID</th>
<th align="left">Time</th>
<th align="left">Tissue</th>
<th align="left">Drug</th>
<th align="left">Patient_ID</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">S1_01</td>
<td align="left">m01</td>
<td align="left">A</td>
<td align="left">d1</td>
<td align="left">p01</td>
</tr>
<tr class="even">
<td align="left">S1_02</td>
<td align="left">m02</td>
<td align="left">A</td>
<td align="left">d1</td>
<td align="left">p01</td>
</tr>
<tr class="odd">
<td align="left">S2_01</td>
<td align="left">m01</td>
<td align="left">A</td>
<td align="left">d1</td>
<td align="left">p02</td>
</tr>
<tr class="even">
<td align="left">S2_02</td>
<td align="left">m02</td>
<td align="left">A</td>
<td align="left">d1</td>
<td align="left">p02</td>
</tr>
<tr class="odd">
<td align="left">S3_01</td>
<td align="left">m01</td>
<td align="left">A</td>
<td align="left">d2</td>
<td align="left">p03</td>
</tr>
<tr class="even">
<td align="left">S3_02</td>
<td align="left">m02</td>
<td align="left">A</td>
<td align="left">d2</td>
<td align="left">p03</td>
</tr>
</tbody>
</table>
<p>In the case above, patients p01 and p02 belong to drug group d1,
while patient p03 belongs to group d2. All patients have samples
collected at months 1 and 2 (encoded as m01 and m02), and all samples
are from the same tissue (A). Note that, in this example, drug group
differentiates samples from different subjects, but not necessarily
samples from the same subject.</p>
<p>The lookup table can also be enriched with other characteristics of
subjects that can be use to partition them, such as age, gender, or
others. Outputs of methods such as clustering can also be added to the
table: this enables exploring, for example, correlations in different
clusters, or comparing trajectories across different clusters.</p>
<div class="section level3">
<h3 id="validation-checks">Validation checks<a class="anchor" aria-label="anchor" href="#validation-checks"></a>
</h3>
<p>The package does a very lightweight validation of the loaded files,
only checking if subjects and samples match. It does not ensure that the
correct transformations have been applied to the expression data, nor
does it warn about or modify missing data – in modules where expression
is paired with a measure, if a subject’s measure is missing that will be
excluded from plots and calculations that include that measure.</p>
<p>The following checks ARE made:</p>
<p><strong>Matching samples and subjects</strong>: the package will
confirm that every sample in the expression matrix is matched to at
least one subject in the lookup table. It will also ensure that all
subjects in the measures table match to at least one sample in the
lookup table. That is, there can be no excess of samples or subjects in
each table.</p>
<p><strong>Matrix format:</strong> if using an .rds file, the package
will check that the expression matrix was indeed saved as a matrix
object in the rds file. This is to ensure that the rownames are read
properly.</p>
</div>
<div class="section level2">
<h2 id="additional-files">Additional files<a class="anchor" aria-label="anchor" href="#additional-files"></a>
</h2>
<div class="section level3">
<h3 id="differential-expression-analysis-results">Differential expression analysis results<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-results"></a>
</h3>
<p>The package includes two modules to showcase results of differential
expression analysis (see <a href="config.html">config</a> for more
details). These modules will read files created using
<code>limma</code>, <code>edgeR</code> or <code>deseq2</code>. All files
should be saved with column names and the column names must not be
changed – the only exception is you want to mix models from different
packages, then you should rename the columns so that all results have
the same column names (e.g. p-values are identified in the same way
across all files).</p>
<p>These modules require the creation of a table that lists all model
results and they support the use of additional columns in the table to
organize results from different types of models or subsets of samples.
All model results file should be placed into a <code>models</code>
folder within the project folder.</p>
<p>The table should look like the following and saved in a CSV or TSV
file:</p>
<table class="table">
<thead><tr class="header">
<th align="left">Model</th>
<th align="left">Time</th>
<th align="left">Drug</th>
<th align="left">File</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Linear</td>
<td align="left">m01</td>
<td align="left">d1</td>
<td align="left">Model_1.txt</td>
</tr>
<tr class="even">
<td align="left">Linear</td>
<td align="left">m02</td>
<td align="left">d2</td>
<td align="left">Model_2.txt</td>
</tr>
<tr class="odd">
<td align="left">Nonlinear</td>
<td align="left">m01</td>
<td align="left">d1</td>
<td align="left">Model_3.txt</td>
</tr>
<tr class="even">
<td align="left">Nonlinear</td>
<td align="left">m02</td>
<td align="left">d2</td>
<td align="left">Model_4.txt</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="gene-moduleslists">Gene modules/lists<a class="anchor" aria-label="anchor" href="#gene-moduleslists"></a>
</h3>
<p>The heatmap module requires the creation of a table containing lists
of names such as gene symbols (see <a href="config.html">config</a> for
more details). In this table, each row will have a column that contains
the gene lists, with symbols separated by a comma. If you have a table
where you have a list identifier and a symbol in each column, you can
use a group-by operation with paste-collapse to create the required
list, as follows:</p>
<div style="display: flex;">
<div>
<p>Original file:</p>
<table class="table">
<thead><tr class="header">
<th align="left">module</th>
<th align="left">gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">A</td>
<td align="left">ABC</td>
</tr>
<tr class="even">
<td align="left">A</td>
<td align="left">DEF</td>
</tr>
<tr class="odd">
<td align="left">A</td>
<td align="left">GHI</td>
</tr>
<tr class="even">
<td align="left">B</td>
<td align="left">JKL</td>
</tr>
<tr class="odd">
<td align="left">B</td>
<td align="left">MNO</td>
</tr>
<tr class="even">
<td align="left">B</td>
<td align="left">PQR</td>
</tr>
</tbody>
</table>
</div>
<div class="col">
<p>  <!-- an empty Div (with a white space), serving as
a column separator --></p>
</div>
<div>
<p>Code to transform:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">table</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">module</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">transmute</span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">gene</span>, collapse <span class="op">=</span> <span class="st">","</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="asking-for-help">Asking for help<a class="anchor" aria-label="anchor" href="#asking-for-help"></a>
</h3>
<p>If you have any issues with data preparation, please post it as an
issue on the package GitHub.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Rafael Henkin.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
